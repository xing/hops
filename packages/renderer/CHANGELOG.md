# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="9.7.0"></a>
# [9.7.0](https://github.com/xing/hops/compare/v9.6.1...v9.7.0) (2018-01-22)




**Note:** Version bump only for package hops-renderer

<a name="9.5.0"></a>
# [9.5.0](https://github.com/xing/hops/compare/v9.4.3...v9.5.0) (2018-01-09)




**Note:** Version bump only for package hops-renderer

<a name="9.4.2"></a>
## [9.4.2](https://github.com/xing/hops/compare/v9.4.1...v9.4.2) (2018-01-08)




**Note:** Version bump only for package hops-renderer

<a name="9.4.1"></a>
## [9.4.1](https://github.com/xing/hops/compare/v9.4.0...v9.4.1) (2018-01-08)




**Note:** Version bump only for package hops-renderer

<a name="9.3.0"></a>
# [9.3.0](https://github.com/xing/hops/compare/v9.2.0...v9.3.0) (2017-12-14)




**Note:** Version bump only for package hops-renderer

<a name="9.1.1"></a>
## [9.1.1](https://github.com/xing/hops/compare/v9.1.0...v9.1.1) (2017-12-05)




**Note:** Version bump only for package hops-renderer

<a name="8.0.0"></a>
# [8.0.0](https://github.com/xing/hops/compare/v7.4.1...v8.0.0) (2017-11-22)


### Features

* **renderer:** add support for an options hash ([8206ad1](https://github.com/xing/hops/commit/8206ad1))
* **renderer:** make renderer use bootstrapServer config ([ce2298d](https://github.com/xing/hops/commit/ce2298d))


### BREAKING CHANGES

* **renderer:** createRenderer signature changed to options hash instead of discrete args




<a name="7.2.0"></a>
# [7.2.0](https://github.com/xing/hops/compare/v7.1.0...v7.2.0) (2017-10-18)




**Note:** Version bump only for package hops-renderer

<a name="7.1.0"></a>
# [7.1.0](https://github.com/xing/hops/compare/v7.0.0...v7.1.0) (2017-10-16)




**Note:** Version bump only for package hops-renderer

<a name="7.0.0"></a>
# [7.0.0](https://github.com/xing/hops/compare/v6.2.8...v7.0.0) (2017-10-13)


### Features

* **renderer:** split hops packages, introduce changelog
