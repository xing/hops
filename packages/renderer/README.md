# Hops Renderer

[![npm](https://img.shields.io/npm/v/hops-renderer.svg)](https://www.npmjs.com/package/hops-renderer)

Hops assumes you will write an Express-style middleware, transpiles it and makes it easy to use in non-transpiled and even non-server code. Hops' renderer is a simple helper to enable you to use your custom middleware outside of Express servers.

Its export, `createRenderer`, creates a `render` function that, if called with a `location` string, returns a promise that resolves to the full body of your middleware's response.

`createRenderer` expects to be passed a `webpackConfig` as first argument. You can also pass `watchOptions` as its second.

### Target Audience

If you want to use a custom Express middleware that's written in ECMAScript Next and transpiled reusing your Webpack loader config, to generate HTML output during your build, you might want to use this renderer. This way, you can prerender some of your pages at buildtime while using the very same code to dynamically generate other pages at runtime.

### Example

This example shows how to write and configure a custom middleware and use it to prerender pages in a build script of some sort.

##### `package.json`

```javascript
{
  ...
  "server": "src/server.js",,
  "dependencies": {
    "hops-build-config": "*",
    "hops-renderer": "*"
  }
  ...
}
```

##### `src/server.js`

```javascript
export default (req, res) => {
  switch (req.url) {
    case '/foo':
      res.write('hello foo');
      break;
    case '/bar':
      res.write('hello bar');
      break;
    default:
      res.writeHead(404);
      res.write('not found');
      break;
  }
  res.end();
};
```

##### `srcipts/render.js`

```javascript
const createRenderer = require('hops-renderer');
const webpackConfig = require('hops-build-config').nodeConfig;

const render = createRenderer(webpackConfig /*, watchOptions */);

render('/foo').then(function(result) {
  // result === 'hello foo'
});

render('/baz').catch(function(error) {
  // error === new Error('invalid status code: 404')
});
```
