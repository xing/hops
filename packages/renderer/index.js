'use strict';

var events = require('events');

var express = require('express');
var mocks = require('node-mocks-http');

var hopsConfig = require('hops-config');
var createMiddleware = require('hops-middleware');

module.exports = function createRenderer(webpackConfig, watchOptions) {
  if (arguments.length === 1 && 'webpackConfig' in webpackConfig) {
    webpackConfig = webpackConfig.webpackConfig;
    watchOptions = webpackConfig.watchOptions;
    console.warn(
      'Calling createRenderer() with just an options hash is deprecated.',
      'Please refer to the latest docs:',
      'https://github.com/xing/hops/blob/master/packages/renderer/README.md'
    );
  }
  var router = new express.Router();
  if (hopsConfig.bootstrapServer) {
    hopsConfig.bootstrapServer(router, hopsConfig);
  }
  router.use(createMiddleware(webpackConfig, watchOptions));
  return function(options) {
    if (typeof options === 'string') {
      options = { url: options };
    }
    return new Promise(function(resolve, reject) {
      var req = mocks.createRequest(options);
      var res = mocks.createResponse({
        eventEmitter: events.EventEmitter,
        request: req,
      });
      res.on('finish', function() {
        if (res.statusCode !== 200) {
          reject(new Error('invalid status code: ' + res.statusCode));
        } else {
          resolve(res._getData());
        }
      });
      router.handle(req, res, function(error) {
        if (error) {
          reject(error);
        } else {
          resolve();
        }
      });
    });
  };
};
