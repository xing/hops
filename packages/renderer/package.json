{
  "name": "hops-renderer",
  "version": "10.0.0-rc.0",
  "description": "Hops HTML renderer based on Webpack",
  "keywords": ["hops", "webpack"],
  "license": "MIT",
  "main": "index.js",
  "files": ["index.js"],
  "dependencies": {
    "hops-config": "10.0.0-rc.0",
    "hops-middleware": "10.0.0-rc.0",
    "node-mocks-http": "^1.6.6"
  }
}
