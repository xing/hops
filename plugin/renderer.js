/**
 * @file plugin/renderer
 *
 * @author Daniel Dembach <daniel@dmbch.net>
 */
'use strict';

var path = require('path');

var webpack = require('webpack');
var MemoryFS = require('memory-fs');

/** @ignore */
exports.eval = function (fileContent, filePath, callback) {
  try {
    Object.keys(require.cache).forEach(function(key) {
      delete require.cache[key];
    });
    var context = Object.assign(
      new module.constructor(),
      { paths: module.paths }
    );
    // eslint-disable-next-line no-underscore-dangle
    context._compile(fileContent.toString(), filePath);
    callback(null, context.exports.default || context.exports);
  }
  catch (error) {
    callback(error);
  }
};

/** @ignore */
exports.createRenderer = function (configFile) {
  if (!configFile) {
    return Promise.resolve(function () {
      return Promise.resolve({ state: '{}', dom: ''});
    });
  }
  return Promise.resolve().then(function () {
    return new Promise(function (resolve, reject) {
      var mfs = new MemoryFS();
      var config = require(configFile);
      var compiler = webpack(config.clean ? config.clean() : config);
      compiler.outputFileSystem = mfs;
      compiler.run(function(compileError) {
        if (compileError) { reject(compileError); }
        else {
          var filePath = path.join(config.output.path, config.output.filename);
          mfs.readFile(filePath, function (readFileError, fileContent) {
            if (readFileError) { reject(readFileError); }
            else {
              exports.eval(fileContent, filePath, function (evalError, result) {
                if (evalError) { reject(evalError); }
                else { resolve(result); }
              });
            }
          });
        }
      });
    });
  });
};
