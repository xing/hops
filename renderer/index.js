'use strict';

var events = require('events');

var mocks = require('node-mocks-http');

var createMiddleware = require('../middleware');

module.exports = function createRenderer (hopsConfig, watchOptions) {
  var middleware = createMiddleware(hopsConfig, watchOptions);

  return function (location) {
    return new Promise(function (resolve, reject) {
      var req = mocks.createRequest({
        url: location
      });
      var res = mocks.createResponse({
        eventEmitter: events.EventEmitter,
        request: req
      });
      res.on('finish', function () {
        if (res.statusCode !== 200) {
          reject('invalid status code: ' + res.statusCode);
        } else {
          resolve(res._getData());
        }
      });
      middleware(req, res, function (error) {
        if (error) { reject(error); } else { resolve(); }
      });
    });
  };
};
