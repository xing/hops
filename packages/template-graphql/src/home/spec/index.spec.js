import React from 'react';
import renderer from 'react-test-renderer';
import { HelmetProvider } from 'react-helmet-async';
import { Home } from '../';

HelmetProvider.canUseDOM = false;

it('renders loading state correctly', () => {
  const tree = renderer
    .create(
      <HelmetProvider>
        <Home data={{ loading: true }} />
      </HelmetProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders loaded state correctly', () => {
  const data = {
    loading: false,
    jobSearchByQuery: {
      collection: [
        {
          jobDetail: {
            id: '1',
            url: 'https://www.xing.com/jobs/1',
            title: 'some job',
            companyInfo: {
              company: {
                companyName: 'cool company',
              },
            },
          },
        },
      ],
    },
  };
  const tree = renderer.create(
    <HelmetProvider>
      <Home data={data} />
    </HelmetProvider>
  );
  expect(tree).toMatchSnapshot();
});
