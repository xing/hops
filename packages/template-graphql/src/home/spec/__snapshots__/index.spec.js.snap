// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders loaded state correctly 1`] = `
<div>
  <h1
    className="headline"
  >
    Hello World!
  </h1>
  <p>
    <a
      href="https://www.xing.com/jobs/1"
    >
      "
      some job
      "
    </a>
     at
     
    <b>
      cool company
    </b>
  </p>
</div>
`;

exports[`renders loading state correctly 1`] = `
<div>
  <h1
    className="headline"
  >
    Hello World!
  </h1>
  <div>
    loading...
  </div>
</div>
`;
